### PR TITLE
Memory Fixes to Higher Order GP Model

### PR DESCRIPTION
**Pull request by wjmaddox**
Detaching the latent dimensions of the HOGP should reduce the memory overhead a bit as we don't need to keep the computation graphs around.

Also fixes a small bug where the full joint covariance was not picking up the computation graph; resolved by cloning the lazy tensor here.

## Motivation

Reduce memory overhead.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/master/CONTRIBUTING.md#pull-requests)?

yes.

## Test Plan

unit tests.

## Related PRs

n/a.

### CODE REVIEW

File: botorch/models/higher_order_gp.py [hunk 1]
```diff
@@ -650,7 +654,7 @@ def posterior(
                 pred_variance = self.make_posterior_variances(joint_covar)
 
                 full_covar = KroneckerProductLazyTensor(
-                    full_covar, *joint_covar.lazy_tensors[1:]
+                    full_covar, *[x.detach() for x in joint_covar.lazy_tensors[1:]]
```
Comment (Balandat) [inline botorch/models/higher_order_gp.py:657]: I guess this will preclude computing gradients of quantities computed on the posterior w.r.t. the latents? I think that's fine since we usually don't do that, but we should probably add a comment here just to be safe.

File: botorch/models/higher_order_gp.py [hunk 2]
```diff
@@ -659,7 +663,7 @@ def posterior(
                 covar = cm(param)
                 if covar.batch_shape != batch_shape:
                     covar = BatchRepeatLazyTensor(covar, batch_shape)
-                joint_covar_list.append(covar)
+                joint_covar_list.append(covar.detach())
```
Comment (Balandat) [inline botorch/models/higher_order_gp.py:666]: why not move this up and do `covar = cm(param).detach()`? That would avoid calling this on a `LazyTensor`

File: botorch/models/higher_order_gp.py (hunk 1 above)
Comment (wjmaddox) [reply botorch/models/higher_order_gp.py:659]: I think it's unlikely we'd need to do that for now, but the updated comment makes it clear.

**Review by Balandat — APPROVED** [review]
